Document infrastructure integration testing decisions (#329)
cloudposse/docs: content/design-decisions/0002-infrastructure-integration-testing.md

---
title: "Strategy for Infrastructure Integration Testing"
date: "2018-12-14" 
description: "We need a consistent way to test and validate that our infrastructure automation (kops, helm, terraform modules, etc) are working correctly."
weight: 2
---

## Status

**Proposed**

## Context

We maintain 250+ active projects on GitHub. These projects receive a lot of contributions from community members. 

We need a way to test `terraform` modules, `kops` clusters, `helm` charts, `helmfile` deployments, and more generally the functionality of our `geodesic` containers.
Our situation is a bit unique in that we need to test the integration of dozens of tools that span languages, authors, and organizations. 
We do not have the privilege of just testing one language or framework.

To date, we have not done much in the form of automated testing. When we receive contributions from members, someone on our team needs to manually checkout the changes to validate them. This is time consuming and we do not have the engineering resources to scale with the growth in popularity of our Open Source projects.

> We have seen approaches to automated testing of terraform.

Author: ```suggestion
We have seen some formal approaches to automated testing of terraform, but nothing for the other tools.
```

### Test Kitchen

> [Test Kitchen](https://github.com/newcontext-oss/kitchen-terraform) is the original testing framework for terraform written in ruby. 

Author: As in capitalization?

Python, Go or Java.

### Terratest

[Terratest](https://github.com/gruntwork-io/terratest) is a project by Gruntworks that they use internally for building and running automated tests against terraform modules. 

The project has received immense interest and is engineering-wise a beautiful solution to the problem. By writing the integration tests in Go and compiling them down to a single, cross-architecture/platform binary, the end-user is able to run tests without needing to install the ["kitchen sink"](https://en.wikipedia.org/wiki/Scope_creep) to testing tools.

The problem with this approach is it sets the bar very high for users of `terraform`, many of which do not come from a strict software development background.
The persons who will most of the time run the tests are not the end users of the modules, but the developers of them. Getting started with Go requires a bit of commitment to get it setup and installed. In the end, we feel that this will exclude a large part of our community, which is contrary to our inclusive ethos. 

> For an example, let us consider the wildly popular blogging platform of Wordpress.Much of its success was derived from the fact that it made it very easy

Reviewer: WordPress

Author: Agree - will fix

> For an example, let us consider the wildly popular blogging platform of Wordpress.Much of its success was derived from the fact that it made it very easy

Reviewer: ```suggestion
For an example, let us consider the wildly popular blogging platform of WordPress. Much of its success was derived from the fact that it made it very easy
```

Bats presents a familiar testing DSL for running adhoc commands using the `bash` shell. This is really nice since it is a general way to test commands 
and outcomes in a largely language/framework agnostic manner. That is, it is not opinionated in the same way as more language-specific testing frameworks. 
Since the tests are just commands wrapped with a simple DSL plus some syntatic sugar, it is easy for anyone who has ever written a shell script to pick it up
in a matter of minutes.

The downside with `bats` is that it is written in `bash`, which is not a *modern* programming language. The underlying code that powers `bats` is complicated
for the uniniated shell programmers. Also, since it is interpreted (not compiled), lots of files/libraries get installed on the system much like a python or ruby app. The tests we write will most likely run commands like `curl` or `jq`, so there are more *native dependencies* that need to be installed. 

> Using the `bats` strategy will probably necessistate running tests in a container to avoid needing to install too many native dependecies. 

Reviewer: ```suggestion
Using the `bats` strategy will probably necessitate running tests in a container to avoid needing to install too many native dependencies. 
```

> We noticed that [HashiCorp](https://www.hashicorp.com/) is using `bats-core` for the automated testing of their [`consul-helm`](https://github.com/hashicorp/consul-helm) chart. This is a good example of the verstatility of `bats` for testing.

Reviewer: ```suggestion
We noticed that [HashiCorp](https://www.hashicorp.com/) is using `bats-core` for the automated testing of their [`consul-helm`](https://github.com/hashicorp/consul-helm) chart. This is a good example of the versatility of `bats` for testing.
```

### Makefile

We could of course use a simple (or not so simple) `Makefile` to run our tests. By default `make` will abort on the first non-zero exit codes recieved by any of the commands. 

The problem with using a general purpose tool like `make` is that it is overloading the purpose of the tool: task running. We would need to develop
our own testing framework and methodology on top of it, which is not our core competency or business differentiator. 

> Also, `make` is not a well understood language by developers because a `Makefile` is basically templated shell scrits, so the execution flow is awkward. 

Reviewer: ```suggestion
Also, `make` is not a well understood language by developers because a `Makefile` is basically templated shell scripts, so the execution flow is awkward. 
```
